### PR TITLE
BZ 1956836: overlay: Add rhcos-usrlocal-selinux-fixup.service

### PR DESCRIPTION
**Pull request by cgwalters**
A fix is inbound for policy, but we really should fixup existing
systems in place.

### CODE REVIEW

**Comment by jlebon** [comment]
Hmm, I think we need this *and* also temporarily carry https://src.fedoraproject.org/rpms/selinux-policy/pull-request/24 until it makes it into RHEL8. Something like:

```diff
diff --git a/manifest.yaml b/manifest.yaml
index 4205ec4..30d74f9 100644
--- a/manifest.yaml
+++ b/manifest.yaml
@@ -163,6 +163,18 @@ postprocess:
      # NB: we don't use -f here so we break when this is no longer needed
      rm -v /etc/iscsi/initiatorname.iscsi

+  # Carry https://src.fedoraproject.org/rpms/selinux-policy/pull-request/24
+  # until it gets into RHEL8. Tracked at https://bugzilla.redhat.com/show_bug.cgi?id=1943381.
+  - |
+    #!/usr/bin/env bash
+    set -xeuo pipefail
+
+    f=/etc/selinux/targeted/contexts/files/file_contexts.subs_dist
+    if ! grep /var/usrlocal $f; then
+      echo "# https://src.fedoraproject.org/rpms/selinux-policy/pull-request/24" >> $f
+      echo "/var/usrlocal /usr/local" >> $f
+    fi
+
 etc-group-members:
   - wheel
   - sudo
```

?

This should also allow you to simplify the systemd service to just do `restorecon -rv /var/usrlocal`. Testing locally (after applying the change above live, not from a new compose):

```
[root@cosa-devsh ~]# echo foobar > /usr/local/sbin/foobar
[root@cosa-devsh ~]# chmod a+x /usr/local/sbin/foobar
[root@cosa-devsh ~]# restorecon -rv /var/usrlocal
Relabeled /var/usrlocal from system_u:object_r:var_t:s0 to system_u:object_r:usr_t:s0
Relabeled /var/usrlocal/etc from system_u:object_r:var_t:s0 to system_u:object_r:usr_t:s0
Relabeled /var/usrlocal/games from system_u:object_r:var_t:s0 to system_u:object_r:usr_t:s0
Relabeled /var/usrlocal/include from system_u:object_r:var_t:s0 to system_u:object_r:usr_t:s0
Relabeled /var/usrlocal/lib from system_u:object_r:var_t:s0 to system_u:object_r:lib_t:s0
Relabeled /var/usrlocal/man from system_u:object_r:var_t:s0 to system_u:object_r:usr_t:s0
Relabeled /var/usrlocal/sbin from system_u:object_r:var_t:s0 to system_u:object_r:bin_t:s0
Relabeled /var/usrlocal/sbin/foobar from unconfined_u:object_r:var_t:s0 to unconfined_u:object_r:bin_t:s0
Relabeled /var/usrlocal/share from system_u:object_r:var_t:s0 to system_u:object_r:usr_t:s0
Relabeled /var/usrlocal/src from system_u:object_r:var_t:s0 to system_u:object_r:usr_t:s0
```

**Comment by travier** [comment]
> This should also allow you to simplify the systemd service to just do `restorecon -rv /var/usrlocal`. Testing locally (after applying the change above live, not from a new compose):

To fix the (SELinux) user part too with `restorecon` you need `-F`

**Comment by cgwalters** [comment]
OK wow I had thought `systemd-tmpfiles` did nothing if the file exists, but it does seem to always reset the security context each boot, plus if e.g. I do `chown bin:bin /usr/local/sbin && reboot` then it gets reset back to `root:root`.  Did you come across this before?

Mmmm.  Locally changing policy feels riskier than just having this service run every time.

**Comment by cgwalters** [comment]
OK reworked to always restorecon `/var/usrlocal/sbin`, but still only do the recursive relabel once.

**Comment by cgwalters** [comment]
The unit test doesn't cover actually injecting binaries via Ignition right now, but I extensively tested this manually too.

File: overlay.d/05rhcos/usr/lib/systemd/system/rhcos-usrlocal-selinux-fixup.service [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[Unit]
+Description=Upgrade fixup: SELinux permissions for /var/usrlocal/sbin
```
Comment (jlebon) [inline overlay.d/05rhcos/usr/lib/systemd/system/rhcos-usrlocal-selinux-fixup.service:2]: Can we link to the RHBZ in here? Maybe even as a `Documentation` key.

Comment (jlebon) [reply overlay.d/05rhcos/usr/lib/systemd/system/rhcos-usrlocal-selinux-fixup.service:2]: Bikeshed:

```suggestion
Description=RHCOS Fix SELinux Labeling For /usr/local/sbin
```

File: tests/kola/misc-ro/misc-ro.sh [hunk 2]
```diff
@@ -133,6 +133,14 @@ echo "ok iSCSI initiator name"
 systemctl is-enabled logrotate.timer
 echo "ok logrotate"
 
+bin_ctx=$(ls -dZ /usr/sbin | cut -f 1 -d ' ')
+usrlocal_sbin_ctx=$(ls -dZ /var/usrlocal/sbin | cut -f 1 -d ' ')
```
Comment (jlebon) [inline tests/kola/misc-ro/misc-ro.sh:137]:
```suggestion
bin_ctx=$(stat -c %C /usr/sbin)
usrlocal_sbin_ctx=$(stat -c %C /var/usrlocal/sbin)
```

File: tests/kola/misc-ro/misc-ro.sh [hunk 3]
```diff
@@ -133,6 +133,14 @@ echo "ok iSCSI initiator name"
...
+if test "${bin_ctx}" != "${usrlocal_sbin_ctx}"; then
+  fatal "/usr/sbin is ${bin_ctx} but /var/usrlocal/sbin is ${usrlocal_sbin_t}"
```
Comment (jlebon) [inline tests/kola/misc-ro/misc-ro.sh:139]:
```suggestion
  fatal "/usr/sbin is ${bin_ctx} but /var/usrlocal/sbin is ${usrlocal_sbin_ctx}"
```

**Review by jlebon — APPROVED** [review]
Minor comments, but LGTM generally!

**Comment by cgwalters** [comment]
Squashed the suggestions (thanks!) and updated `Documentation=`

**Comment by travier** [comment]
> OK wow I had thought `systemd-tmpfiles` did nothing if the file exists, but it does seem to always reset the security context each boot, plus if e.g. I do `chown bin:bin /usr/local/sbin && reboot` then it gets reset back to `root:root`. Did you come across this before?

I had a similar issue earlier: https://github.com/coreos/ignition/issues/1156

File: tests/kola/misc-ro/misc-ro.sh [hunk 4]
```diff
@@ -133,6 +133,21 @@ echo "ok iSCSI initiator name"
 systemctl is-enabled logrotate.timer
 echo "ok logrotate"
 
+<<<<<<< HEAD
```
Comment (travier) [inline tests/kola/misc-ro/misc-ro.sh:136]: Looks like a broken merge 🙂

**Comment by jlebon** [comment]
>  INFO[2021-05-19T16:25:48Z] error: Packages not found: audit, authselect, bash-completion, bind-utils, bsdtar, bzip2, chrony, cifs-utils, cloud-utils-growpart, compat-openssl10, conntrack-tools, coreutils, cryptsetup, cryptsetup-reencrypt, device-mapper-multipath, dhcp-client, dnsmasq, e2fsprogs, efibootmgr, git-core, glusterfs-fuse, gnupg2, grub2, grub2-efi-x64, gzip, hostname, iproute, iproute-tc, iptables, irqbalance, iscsi-initiator-utils, kernel, kernel-core, kernel-modules, kernel-modules-extra, kexec-tools, less, logrotate, lvm2, mdadm, microcode_ctl, nano, net-tools, nfs-utils, nftables, nmap-ncat, open-vm-tools, openssh-clients, openssh-server, passwd, perl-interpreter, qemu-guest-agent, rdma-core, rpm-ostree, rsync, selinux-policy-targeted, sg3_utils, shadow-utils, shim, socat, sssd, stalld, strace, subscription-manager-rhsm-certificates, sudo, systemd, systemd-journal-gateway, systemd-journal-remote, tar, teamd, tmux, tpm2-tools, vim-minimal, xfsprogs, xz 

Hmm, something to do with the recent repo changes I suppose? /cc @travier

**Comment by jlebon** [comment]
/lgtm
for PR itself!

**Comment by cgwalters** [comment]
/hold 
until repos are fixed

**Comment by travier** [comment]
Should be good to go now but maybe a full retest would be best

**Pull request merged**